### PR TITLE
Add `HammingWeightCompute` gate using 3-to-2 Gidney adders

### PR DESCRIPTION
**Pull request by tanujkhattar**
Follows construction described on Page-4 of https://arxiv.org/abs/1709.06648

cc @ncrubin

### CODE REVIEW

File: qualtran/bloqs/arithmetic.py [hunk 1]
```diff
@@ -578,6 +578,109 @@ def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
         return {(Add(self.bitsize), 1), (ArbitraryClifford(n=2), self.bitsize)}
 
 
+@frozen
+class HammingWeightCompute(GateWithRegisters):
+    r"""A gate to compute the hamming weight of n-bit register in a new log_{n} bit register.
```
Comment (fdmalone) [inline qualtran/bloqs/arithmetic.py:583]: nit: "an n-bit register"

File: qualtran/bloqs/arithmetic.py [hunk 2]
```diff
@@ -578,6 +578,109 @@ def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
...
+
+    Implements $U|x\rangle |0\rangle \rightarrow |x\rangle|hamming_weight(x)\rangle$
```
Comment (fdmalone) [inline qualtran/bloqs/arithmetic.py:585]: probably want some \mathrm's around the hamming_weight

File: qualtran/bloqs/arithmetic.py [hunk 3]
```diff
@@ -578,6 +578,109 @@ def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
...
+        bitsize: Number of bits in the input register. The allocated output register
+            is of size `log_2(bitsize)` so it has enough space to hold the hamming weight of x.
+
+    Registers:
+     - x: A bitsize-sized input register (register a above).
+     - junk: A temporary ancilla register of size bitsize.
+     - out: A log2(bitize)-sized LEFT/RIGHT register depending on whether the gate adjoint or not.
+
+    References:
+        [Halving the cost of quantum addition](https://arxiv.org/abs/1709.06648)
```
Comment (fdmalone) [inline qualtran/bloqs/arithmetic.py:598]: add a page number ~ rough location if relevant.

File: qualtran/bloqs/arithmetic.py [hunk 4]
```diff
@@ -578,6 +578,109 @@ def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
...
+        return Signature(
+            [
+                Register('x', self.bitsize),
+                Register('junk', sum(self._junk_sizes), side=side),
+                Register('out', self.bitsize.bit_length(), side=side),
+            ]
+        )
+
+    @cached_property
+    def _junk_sizes(self):
```
Comment (fdmalone) [inline qualtran/bloqs/arithmetic.py:616]: return type -> List[int]?

File: qualtran/bloqs/arithmetic.py [hunk 5]
```diff
@@ -578,6 +578,109 @@ def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
...
+        bitsize = self.bitsize
+        ret = []
+        while bitsize > 0:
+            bitsize = bitsize // 2
+            ret.append(bitsize)
+        return ret
+
+    def short_name(self) -> str:
+        return "out = x.bit_count()"
```
Comment (fdmalone) [inline qualtran/bloqs/arithmetic.py:625]: nice

File: qualtran/bloqs/arithmetic.py [hunk 6]
```diff
@@ -578,6 +578,109 @@ def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
...
+            else:
+                anc = junk.pop()
+                yield self._three_to_two_adder(x[-2], x[-1], out[out_idx], anc)
+                y.append(anc)
+            x = [*y]
+
+    def decompose_from_registers(
+        self, *, context: cirq.DecompositionContext, **quregs: NDArray[cirq.Qid]
+    ) -> cirq.OP_TREE:
+        x: List[cirq.Qid] = [*quregs['x'][::-1]]
```
Comment (fdmalone) [inline qualtran/bloqs/arithmetic.py:655]: Why are we reversing the qubit order? worth a note I think.

**Review by fdmalone — CHANGES_REQUESTED** [review]
LGTM % nits

File: qualtran/bloqs/arithmetic.py [hunk 7]
```diff
@@ -578,6 +578,109 @@ def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
         return {(Add(self.bitsize), 1), (ArbitraryClifford(n=2), self.bitsize)}
 
 
+@frozen
+class HammingWeightCompute(GateWithRegisters):
+    r"""A gate to compute the hamming weight of n-bit register in a new log_{n} bit register.
+
+    Implements $U|x\rangle |0\rangle \rightarrow |x\rangle|hamming_weight(x)\rangle$
+    using $4nT$ gates. Uncomputation requires 0 T-gates.
```
Comment (ncrubin) [inline qualtran/bloqs/arithmetic.py:586]: Ian's paper and the Hubbard paper both say n-1 toffoli  But Craig's paper says n Toffoli.  Is there an accounting error? or a reporting error?

**Review by ncrubin — APPROVED** [review]

File: qualtran/bloqs/arithmetic.py (hunk 4 above)
Comment (tanujkhattar) [reply qualtran/bloqs/arithmetic.py:616]: removed `junk_sizes`

File: qualtran/bloqs/arithmetic.py (hunk 2 above)
Comment (fdmalone) [reply qualtran/bloqs/arithmetic.py:585]: And also check that the `_` renders properly sometimes this has a weird effect

File: qualtran/bloqs/arithmetic.py (hunk 6 above)
Comment (tanujkhattar) [reply qualtran/bloqs/arithmetic.py:697]: Added a note

File: qualtran/bloqs/arithmetic.py (hunk 2 above)
Comment (tanujkhattar) [reply qualtran/bloqs/arithmetic.py]: Done

**Comment by tanujkhattar** [comment]
@fdmalone Addressed all comments, improved the bounds on Toffoli's and ancilla's in the docstrings and simplified computation of `junk_bitsize`. I think this is good to go now.

**Review by fdmalone — APPROVED** [review]

**Pull request merged**